### ISSUE TITLE
WO-009: step function and env wrapper

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#14]
> **WO-009** - Phase 1 - Owner **LEAD** - Difficulty **5/5** - Gate **G0**
> Card: [`workorders/WO-009.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-009.md)

> ## LEAD-OWNED CARD - AN IMPLEMENTER SESSION MUST NOT PICK THIS UP
> PLAN sections 12.3 and 1.3 (finding F14) name the step function, the JAX port and the PPO adapter
> as the three units the lead writes itself. The step function is where the information invariants of
> CONTRACT rules 5, 6 and 9 are either preserved or quietly broken, and every `gosplan/env/` diff is
> reviewed against rule 7 before gate G0 is signed. **An implementer session that finds itself editing
> `gosplan/env/state.py`, `gosplan/env/step.py` or `gosplan/env/env.py` has taken the wrong card.**

## What this is

This card assembles the whole environment. It writes the period schedule as an explicit, auditable
state machine - eight typed stages, DELIVER through TERMINATE - the state record and its lifecycle,
and the `GosplanEnv` reset/step wrapper the agents and experiments drive. Every stage function is a
thin assembler that calls the module owning the arithmetic and threads `State` through; no formula
from PLAN sections 2.6-2.11 may be restated here.

It exists because the ordering is the design. TARGET runs after REWARD so this period's bonus is
judged against this period's own target; the holding loss lands before the period's output is added;
the horizon is one global geometric draw and `done` is the only channel by which it reaches the
learner. Get the ordering wrong and the golden files disagree with the reference - which is exactly
what T-B7 is for.

## Depends on

- WO-005
- WO-006
- WO-007
- WO-008

The fan PLAN section 12.2 puts ahead of this card - and through them WO-001 to WO-004. The frozen
suite is WO-002's.

**Blocked by:** #10 (WO-005), #11 (WO-006), #12 (WO-007), #13 (WO-008)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- `PLAN.md` - **not in the repository**: git-ignored, supplied out of band, required at the
  repository root. Sections 2.2 (state), 2.3 (actions), 2.4 (observation), 2.5 (period schedule - the
  core of this card), 2.11 (inventory), 2.12 (horizon), 2.15 (RNG), 8 (layout), 11 (test
  architecture), 12.3 (this card), 13 (gate G0). Sections 2.6 to 2.10 for **ordering only** - their
  formulas belong to the modules that own them.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rules 4, 5, 6, 7, 8,
  9, 10 in full.
- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - the frozen
  `State`, `EnterpriseAction`, `StepInfo`, `action_spec` and `active_action_dims` declarations.
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py) -
  `EnvConfig` and every field the schedule reads.
- [`gosplan/params.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/params.py) - the
  registry behind those defaults.
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py) - `draw`,
  `PURPOSES`; purpose `terminate` is this card's.
- [`gosplan/env/state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py) -
  **write target.** `State`, `EnterpriseAction`, `StepInfo` field declarations plus four function
  contracts.
- [`gosplan/env/step.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/step.py) -
  **write target.** `PeriodStage`, `PERIOD_SCHEDULE`, `AGENT_STAGES` are already there as data; twelve
  bodies missing.
- [`gosplan/env/env.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/env.py) -
  **write target.** `action_spec`, `active_action_dims`, `GosplanEnv`.
- [`gosplan/env/production.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/production.py) -
  `coverage`, `produce_step` (WO-005's arithmetic).
- [`gosplan/env/planner.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/planner.py) -
  `make_planner_view`, `update_targets`, `fulfilment_measure`, `allocate`, `deliver`, `select_audits`
  (WO-006's).
- [`gosplan/env/reporting.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reporting.py) -
  `process_reports`, `audit_and_penalise` (WO-007's).
- [`gosplan/env/reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reward.py) -
  `bonus`, `reward_scale`, `enterprise_reward`, `val_measured`, `val_true`, `welfare_true` (WO-007's).
- [`gosplan/env/obs.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/obs.py) -
  `obs_spec`, `build_observation`, `phase_mask` (WO-008's).
- [`gosplan/env/prices.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/prices.py) -
  `initial_prices`, precomputed once at construction (WO-007's).
- [`gosplan/env/trade.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/trade.py) -
  the Phase-2 sketch `stage_trade` routes through; inert at P1.
- [`gosplan/env/ministry.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/ministry.py) -
  the Phase-2 ministry layer; inert at P1.
- [`gosplan/env/__init__.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/__init__.py) -
  the package surface.
- [`gosplan/metrics/ledger.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/metrics/ledger.py) -
  `StepRecord`, `Ledger`, `MANIFEST_FIELDS` - what `StepInfo` feeds.
- `ref/ref_step.py` - the reference dynamics the golden files come from (written by WO-002).
- `ref/gen_golden.py` - the generator and its invocation contract (written by WO-002).
- `docs/ref_worked_example.md` - the hand-checked two-enterprise case (written by WO-002).
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - ruff
  line-length 100, target py312; `testpaths`.
- [`Makefile`](https://github.com/Xocas12/gosplan-env/blob/main/Makefile) - the targets the completion
  command mirrors.
- `tests/unit/test_conservation.py` - read-only (rule 2), from WO-002.
- `tests/unit/test_env_api.py` - read-only (rule 2), from WO-002.
- `tests/behavioural/test_termination.py` - read-only (rule 2), from WO-002.
- `tests/golden/test_golden_parity.py` - read-only (rule 2), from WO-002.
- `tests/golden/*.json` - generated artefacts; read-only here.

Nothing else. Not `gosplan/agents/`, not `gosplan/experiments/`, not `runs/`.

Read nothing else. CONTRACT rule 12.

## Write only

- `gosplan/env/state.py` - `State`, `EnterpriseAction`, `StepInfo`, `initial_targets`,
  `initial_state`, `reset_period_accumulators`, `advance_phase`
- `gosplan/env/step.py` - `stages_for_step`, `stage_deliver`, `stage_trade`, `stage_produce`,
  `stage_report`, `stage_audit`, `stage_reward`, `stage_target`, `stage_terminate`, `advance`,
  `run_period`
- `gosplan/env/env.py` - `action_spec`, `active_action_dims`, `GosplanEnv`

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

- [ ] `tests/golden/test_golden_parity.py` - **T-B7**: implementation equals `ref/ref_step.py` to
      1e-9 on every seeded trajectory (5 configs x 3 seeds x 2 agents).
- [ ] `tests/unit/test_conservation.py` - **T-U1**: per-period, per-good conservation to 1e-9:
      `sum y + sum S_prev = sum inputs consumed / a + sum consumer + sum S_next + holding loss + cap overflow`.
- [ ] `tests/unit/test_env_api.py` - reset/step contract, the three `stages_for_step` cases,
      `obs_spec` / `action_spec` / `active_action_dims` agreement, and **T-U5** permutation
      invariance.
- [ ] `tests/behavioural/test_termination.py` - **T-B9**: empirical continuation equals `psi`; no
      observation field correlates with periods remaining (coefficient ~ 0).

**No regression** in what is already green: `tests/unit/test_config.py`, `test_rng.py`,
`test_production.py`, `test_planner.py`, `test_reporting.py`, `test_reward.py`, `test_prices.py`,
`test_obs.py`, `tests/behavioural/test_planner_blindness.py`, `test_welfare_blindness.py`. They are
in the completion command for that reason. T-B1, T-B2 and T-B3 need WO-010's heuristics and are NOT
on this card's list.

```bash
uv run ruff check gosplan/env/state.py gosplan/env/step.py gosplan/env/env.py
uv run ruff format --check gosplan/env/state.py gosplan/env/step.py gosplan/env/env.py
uv run python -m ref.gen_golden
uv run pytest tests/unit/test_conservation.py tests/unit/test_env_api.py tests/behavioural/test_termination.py tests/golden/test_golden_parity.py -q
uv run pytest tests/unit/test_config.py tests/unit/test_rng.py tests/unit/test_production.py tests/unit/test_planner.py tests/unit/test_reporting.py tests/unit/test_reward.py tests/unit/test_prices.py tests/unit/test_obs.py tests/behavioural/test_planner_blindness.py tests/behavioural/test_welfare_blindness.py -q
```

The third line runs **only** when `tests/golden/*.json` is absent (the files are git-ignored). It is
never run to clear a parity failure.

### 1. Section 2.5 as an explicit state machine

This is the defining requirement of the card. The eight stages are **data, not control flow**:
`PeriodStage` (DELIVER 0, TRADE 1, PRODUCE 2, REPORT 3, AUDIT 4, REWARD 5, TARGET 6, TERMINATE 7),
`PERIOD_SCHEDULE` as the ordered tuple, `AGENT_STAGES = (PRODUCE, REPORT)`, and
`stages_for_step(state, cfg)` as the transition function returning a **contiguous slice** of
`PERIOD_SCHEDULE`, with `M = cfg.incentive.steps_per_period`:

```
k_step == 0,    phase "produce"   ->  (DELIVER, TRADE, PRODUCE)
0 < k_step < M, phase "produce"   ->  (PRODUCE,)
k_step == M,    phase "report"    ->  (REPORT, AUDIT, REWARD, TARGET, TERMINATE)
```

`advance` executes exactly the stages `stages_for_step` returns, in that order, and nothing else. The
schedule must be auditable by reading data, never inferred from the body of a step function - PLAN
section 2.5 is titled "period schedule (typed)" for that reason, and finding F10 records the typed
schedule with a separate REPORT step as the fix for the original sequencing ambiguity. `stage.value`
orders the schedule and `PERIOD_SCHEDULE[stage.value] is stage`.

### 2. Agent-steps versus stages

The agent acts `M + 1` times per period. DELIVER opens the period at the head of its first PRODUCE
step and consumes the claims recorded at the **previous** period's REPORT step. TRADE sits between
DELIVER and PRODUCE because it reallocates the inputs DELIVER has just placed in `X`. AUDIT, REWARD,
TARGET and TERMINATE follow the REPORT step with no further agent input.

### 3. Each stage function is a thin assembler

It calls the module that owns the arithmetic and threads `State` through. **No formula of PLAN
sections 2.6 to 2.11 may be restated in these three files** - production is `production.py`, planner
rules and physical delivery `planner.py`, reporting/audit/penalty/bonus/reward/val/welfare
`reporting.py` and `reward.py`, the observation `obs.py`, prices `prices.py`. Duplicating a formula
here creates a second place for T-B7 to disagree with `ref/ref_step.py`.

### 4. Ordering that is load-bearing

TARGET runs **after** REWARD, so this period's bonus is judged against this period's own target (PLAN
section 2.7.1). The holding loss is applied before the period's output is added at REPORT
(`S <- (1 - h) * S + y`, sections 2.8 and 2.11) - inside `reporting.py`, called from `stage_report`.
TERMINATE is last.

### 5. Horizon (section 2.12)

Under `horizon_mode = "geometric"` the episode runs `cfg.tech.min_periods` periods unconditionally,
then continues with probability `cfg.incentive.tenure` per period, hard cap `cfg.tech.max_periods`.
**One global draw** with purpose `terminate`, not a per-enterprise flag. `done` is the only channel by
which the outcome reaches the learner; **no observation field carries periods remaining** and none may
proxy it (finding F4, T-B9). `tenure` is the economic INC parameter and is never conflated with the
PPO discount `gamma`, which lives with the adapter (WO-017) and which the environment never reads.

### 6. RNG (rule 9)

Every stochastic term goes through `gosplan.rng.draw(seed_env, purpose, *indices)` with a purpose from
the frozen `Purpose` literal. No `numpy.random`, no `jax.random`, no module-level generator anywhere
under `gosplan/env/`. `seed_policy` is carried on the state as a separate stream and is never mixed
into an environment draw.

### 7. `StepInfo` is not agent-facing (rule 6)

It carries `y`, `S`, `X`, the audit measurement and the period-level `val_measured`, `val_true` and
`welfare`; it is written by the environment and read by `gosplan/metrics/ledger.py` and lead-run
experiments only. `GosplanEnv.step` returns it **beside** the observation, never inside it. The three
period-level scalars are filled at the point of the schedule that defines them and are zero
elsewhere: `val_measured` and `val_true` at REPORT, `welfare` after DELIVER, when `consumer` is known.

### 8. Reward is a vector `(N,)`

With exactly the rule-4 terms: `-scale * c_ik` at a PRODUCE step,
`scale * (B(rho) - 1[audited] * Pen + trade_surplus)` at the REPORT step, with
`scale = reward_scale(cfg)` computed once analytically. No shaping, no auxiliary term, no running
normalisation, no per-step bonus of any other kind.

### 9. Actions

`action_spec(cfg)` returns all six dimensions with the PLAN section 2.3 bounds;
`active_action_dims(cfg)` returns `["effort", "report_ratio", "input_request"]` at
`p1_default_config()` and must be a pure function of the configuration that does not change within a
run. Dimensions irrelevant to the current stage are **ignored, never rejected**, and the environment
never trusts an agent to have masked them itself. `input_request` arrives as a multiple of need and is
rescaled and clipped against the current `need_ij` when read.

### 10. Golden files

`tests/golden/*.json` is git-ignored and generated from the frozen reference. If the directory holds
only `.gitkeep`, generate the set once with the unmodified generator (`uv run python -m
ref.gen_golden`, no arguments - the pre-registered matrix, 5 configs x 3 seeds x 2 agents = 30 files,
30 agent-steps each).

**NEVER REGENERATE THEM TO MAKE A FAILING PARITY TEST PASS**: a mismatch is a bug in `gosplan/env/`,
and WO-013 is the only card that regenerates after a deliberate dynamics change. Never `--force`,
never a restricted filter.

## Forbidden

- Restating any formula of PLAN sections 2.6 to 2.11 inside `step.py`, `env.py` or `state.py`.
- Any `numpy.random` or `jax.random` call, or any module-level global generator, under `gosplan/env/`
  (rule 9).
- Any agent-facing use of `StepInfo`; any true quantity reaching an observation or a reward (rules 5,
  6). `make_planner_view` and `build_observation` remain the only sanctioned readers of `State` on
  someone else's behalf.
- Adding, shaping, normalising or rescaling a reward term; any running reward normalisation (rule 4).
- Implementing bunching, padding, storming, hoarding, shaving or trade as a transition rule (rule 7).
  The stages call rules; the phenomena are consequences of those rules or they are not results at all.
- Widening or narrowing `report_max_ratio`, the inventory cap, or any other bound to make a test pass
  (rule 8).
- Reordering `PERIOD_SCHEDULE`, or changing `PeriodStage` values, without a lead decision - it changes
  the dynamics and invalidates every golden file.
- Editing `ref/`, `spec/spec.py`, any test file, `PLAN.md`, or any file not on the "Write only" list,
  including "obvious" one-line fixes in `production.py` or `planner.py`. A defect found there is an
  AMBIGUITY REPORT naming the owning work order.
- Running anything in `tests/acceptance/` (rule 13).

## Contract rules that bite here

- **Rule 3 - STOP AND REPORT.** Choosing "the reasonable default" is a violation: do not invent a
  stage ordering PLAN section 2.5 does not state, and do not resolve a golden-file mismatch by
  adjusting the dynamics until it goes away.
- **Rule 12 - WORK ORDERS (whitelist).** Read only the list above, write only the three files, run
  the completion command verbatim.
- **Rule 4 - REWARD TERMS.** The vector reward carries exactly the named terms, with an analytic
  `scale`; no shaping, no running normalisation.
- **Rule 5 - PLANNER BLINDNESS.** The step machine must not hand a planner rule anything but a
  `PlannerView`.
- **Rule 6 - WELFARE BLINDNESS.** `StepInfo` is returned beside the observation, never inside it, and
  is read only by the ledger and lead-run experiments.
- **Rule 7 - NO HARD-CODED PATHOLOGY.** The stages call rules; every `gosplan/env/` diff is reviewed
  against this rule before G0 is signed.
- **Rule 8 - BOUNDS ARE RESULTS.** No bound is moved to make a test pass.
- **Rule 9 - RNG.** Every draw goes through `gosplan.rng.draw`; `seed_policy` never mixes into an
  environment key.
- **Rule 10 - MANIFEST.** The environment's seeds and configuration hash are what the manifest
  records; nothing here may make a run irreproducible.
- **Rule 2 - FROZEN TESTS.** Every test file and every golden `.json` is read-only here.
- **Rule 13 - TESTS ARE NOT EXPERIMENTS.** Nothing in `tests/acceptance/` is run.

## Definition of done

- [ ] T-B7, T-U1, `tests/unit/test_env_api.py` (incl. T-U5) and T-B9 green, with the number of golden
      files actually compared stated.
- [ ] The no-regression run green.
- [ ] `uv run ruff check . && uv run ruff format --check .` clean.
- [ ] `uv run python scripts/contract_guard.py --all` clean.
- [ ] No file outside "Write only" touched.
- [ ] CI `ci-ok` green on the PR.
- [ ] The four-part report posted as the PR body: 1) tests passed/failed with output - the four
      must-pass items and the no-regression run, plus the number of golden files compared;
      2) ambiguity reports filed, each listed, or "none"; 3) files changed - expect exactly
      `gosplan/env/state.py`, `gosplan/env/step.py`, `gosplan/env/env.py`; 4) anything you were
      tempted to invent and did not - every place the stage ordering, a bound or an RNG key was
      under-determined and you stopped instead of choosing, and an explicit statement of whether any
      golden file was regenerated, and why.
